\documentclass[fleqn,10pt]{wlscirep}
\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{float}
\usepackage{lineno}
\usepackage{graphicx}
\linenumbers

\title{\emph{Skew}DB: A comprehensive database of GC and 10 other skews for over 30,000 chromosomes and plasmids}
% CHANGE: New title, now 30,000

\author[*]{Bert Hubert}
\affil[*]{corresponding author: Bert Hubert (bert@hubertnet.nl)}

\begin{abstract}
  GC skew denotes the relative excess of G nucleotides over C nucleotides on the leading versus the lagging replication strand of eubacteria. While the effect is small, typically around 2.5\%, it is robust and pervasive. GC skew and the analogous TA skew are a localized deviation from Chargaff's second parity rule, which states that G and C, and T and A occur with (mostly) equal frequency even within a strand.

  Different bacterial phyla show different kinds of skew, and differing relations between TA and GC skew.

% CHANGE: Update to 30,000  
  This article introduces an open access database (\url{https://skewdb.org}) of GC and 10 other skews for over 30,000 chromosomes and plasmids. Further details like codon bias, strand bias, strand lengths and taxonomic data are also included. The \emph{Skew}DB can be used to generate or verify hypotheses. Since the origins of both the second parity rule and GC skew itself are not yet satisfactorily explained, such a database may enhance our understanding of prokaryotic DNA.
\end{abstract}
\begin{document}

\flushbottom
\maketitle
\thispagestyle{empty}

\section*{Background \& Summary}
% CHANGE: Added three references that introduce GC skew
The phenomenon of GC skew\cite{pmid:10570985, Marn2008, Quan2019} is tantalizing because it enables a simple numerical analysis that accurately predicts the loci of both the origin and terminus of replication in most bacteria and some archaea \cite{lu_skewit_2020,luo_doric_2019}.

Bacterial DNA is typically replicated simultaneously on both strands, starting at the origin of replication \cite{ODonnell2013}. Both replication forks travel in the 5' to 3' direction, but given that the replichores are on opposite strands, topologically they are traveling in opposite directions. This continues until the forks meet again at the terminus. This means that roughly one half of every strand is replicated in the opposite direction of the other half. The forward direction is called the leading strand. Many plasmids also replicate in this way \cite{Lilly2015}.

The excess of G over C on the leading strand is in itself only remarkable because of Chargaff's somewhat mysterious second parity rule\cite{rudner_separation_1968}, which states that within a DNA strand, there are nearly equal numbers of G's and C's, and similarly, T's and A's. This rule does not directly follow from the first parity rule, which is a simple statement of base pairing rules.

Depending on who is asked, Chargaff's second parity rule is so trivial that it needs no explanation, or it requires complex mathematics and entropy considerations to explain its existence \cite{fariselli_dna_2020}.

The origins of GC skew are still being debated. The leading and lagging strands of circular bacterial chromosomes are replicated very differently; it is at least plausible that this leads to different mutational biases. In addition, there are selection biases that are theorized to be involved \cite{tillier_contributions_2000}. No single mechanism may be exclusively responsible.

This article does not attempt to explain or further mystify \cite{zhang_brief_2014} the second parity rule or GC skew, but it may be that the contents of the \emph{Skew}DB can contribute to our further understanding.

The \emph{Skew}DB also contains some hard to explain data on many chromosomes. These include highly asymmetric skew, but also very disparate strand lengths. Conversely, the \emph{Skew}DB confirms earlier work on skews in the Firmicute phylum  \cite{charneski_atypical_2011}, and also expands on these earlier findings.

% CHANGE: Added the classical GC Skew formula and how it relates to cumulative skew
GC skew has often been investigated by looking at windows of DNA of a certain size. GC skew is computed as $(G-C)/(G+C)$ in a window of $N$ bases, where $G$ is the number of guanines and $C$ the number of cytosines in that window. It has been found that the choice of window size impacts the results of the analysis. The \emph{SkewDB} is therefore based exclusively on cumulative skew \cite{grigoriev_analyzing_1998}, which sidesteps window size issues. For example, the sequence GGGCCC has a cumulative GC skew of zero, and as we progress through the sequence, this skew takes on values 1, 2, 3, 2, 1, 0. If the window size $N$ is 6, the non-cumulative skew is also 0. 

The result of a cumulative GC skew analysis is shown in figure \ref{fig:explainer-graph}. The analysis software fits a linear model on the skews, where it also compensates for chromosome files sequenced in the non-canonical direction, or where the origin of replication is not at the start of the file.

GC skew analysis is not new. As noted below, the DoriC database for example contains related data that is more precise for its stated purpose (finding the Origin of replication). The SkewIT database \cite{lu_skewit_2020} similarly provides a metric of skew, and also comes with an online analysis tool.

Other work, like the Comparative Genometrics Database\cite{Roten2002} and the Z Curve Database \cite{Zhang2003} has been foundational, but by dint of their age lack an analysis of the tens of thousands of DNA sequences that have become available since the initial availability of these databases.

\emph{Skew}DB is funded to be updated monthly with the latest sequences from NCBI until 2026.

Other software that calculates GC skew is available, like for example GraphDNA\cite{Thomas2007}, GC Skewing\cite{Grigoriev1998} and GenSkew. The \emph{Skew}DB delivers far more metrics however, also because it involves annotation data in its calculations. For ease of use, \emph{Skew}DB is made available as a ready to use database, as well as in software form that reproduces this database exactly.

\section*{Methods}
The \emph{Skew}DB analysis relies exclusively on the tens of thousands of FASTA and GFF3 files available through the NCBI download service, which covers both GenBank and RefSeq. The database includes bacteria, archaea and their plasmids.

Furthermore, to ease analysis, the NCBI Taxonomy database is sourced and merged so output data can quickly be related to (super)phyla or specific species.

No other data is used, which greatly simplifies processing. Data is read directly in the compressed format provided by NCBI. All results are emitted as standard CSV files.

In the first step of the analysis, for each organism the FASTA sequence and the GFF3 annotation file are parsed. Every chromosome in the FASTA file is traversed from beginning to end, while a running total is kept for cumulative GC and TA skew. In addition, within protein coding genes, such totals are also kept separately for these skews on the first, second and third codon position. Furthermore, separate totals are kept for regions which do not code for proteins.

In addition, to enable strand bias measurements, a cumulative count is maintained of nucleotides that are part of a positive or negative sense gene. The counter is increased for positive sense nucleotides, decreased for negative sense nucleotides, and left alone for non-genic regions. A separate counter is kept for non-genic nucleotides. 

Finally, G and C nucleotides are counted, regardless of if they are part of a gene or not.

These running totals are emitted at 4096 nucleotide intervals, a resolution suitable for determining skews and shifts.

% CHANGE: Fixed a typo
In addition, one line summaries are stored for each chromosome. These lines include the RefSeq identifier of the chromosome, the full name mentioned in the FASTA file, plus counts of A, C, G and T nucleotides. Finally five levels of taxonomic data are stored.

Chromosomes and plasmids of fewer than 100 thousand nucleotides are ignored, as these are too noisy to model faithfully. Plasmids are clearly marked in the database, enabling researchers to focus on chromosomes if so desired.

\subsection*{Fitting}
Once the genomes have been summarised at 4096-nucleotide resolution, the skews are fitted to a simple model.

\begin{figure}[ht]
\centering
\includegraphics[width=0.7\linewidth]{explainer.pdf}
% CHANGE: Added italics
\caption{Sample graph showing \emph{Skew}DB data for \emph{Lactiplantibacillus plantarum} strain LZ95 chromosome}
\label{fig:explainer-graph}
\end{figure}

The fits are based on four parameters, as shown in figure \ref{fig:explainer-graph}. {\tt Alpha1} and {\tt alpha2} denote the relative excess of G over C on the leading and lagging strands. If {\tt alpha1} is $0.046$, this means that for every 1000 nucleotides on the leading strand, the cumulative count of G excess increases by 46.

The third parameter is {\tt div} and it describes how the chromosome is divided over leading and lagging strands. If this number is $0.557$, the leading replication strand is modeled to make up $55.7\%$ of the chromosome.

The final parameter is {\tt shift} (the dotted vertical line), and denotes the offset of the origin of replication compared to the DNA FASTA file. This parameter has no biological meaning of  itself, and is an artifact of the DNA assembly process. 

The goodness-of-fit number consists of the root mean squared error of the fit, divided by the absolute mean skew. This latter correction is made to not penalize good fits for bacteria showing significant skew.

GC skew tends to be defined very strongly, and it is therefore used to pick the {\tt div} and {\tt shift} parameters of the DNA sequence, which are then kept as a fixed constraint for all the other skews, which might not be present as clearly.

The fitting process itself is a downhill simplex method optimization\cite{Nelder1965} over the four dimensions, seeded with the average observed skew over the whole genome, and assuming there is no shift, and that the leading and lagging strands are evenly distributed. To ensure that the globally optimum fit is (very likely) achieved, ten optimization attempts are made from different starting points. This fitting process is remarkably robust in the sense that even significant changes in parameters or fitting strategies cause no appreciable change in the results. 

% CHANGE: Fixed a broken sentence
For every chromosome and plasmid the model parameters are stored, plus the adjusted root mean squared error.

Both for quality assurance and ease of plotting, individual CSV files are generated for each chromosome, again at 4096 nucleotide resolution, but this time containing both the actual counts of skews as well as the fitted result.
  
\subsection*{Some sample findings}
To popularize the database, an online viewer is available on \url{https://skewdb.org/view}. While this article makes no independent claims to new biological discoveries, the following sections show some results gathered from a brief study of the database. Some of these observations may be of interest for other researchers.
\subsubsection*{GC and TA skews}
Most bacteria show concordant GC and TA skew, with an excess of G correlating with an excess of T. This does not need to be the case however. Figure \ref{fig:gc-ta-scatter} is a scatterplot that shows the correlation between the skews for various major superphyla. Firmicutes (part of the Terrabacteria group) show clearly discordant skews.

\begin{figure}[ht]
\centering
\includegraphics[width=.8\linewidth]{phylo-histo.png}
\caption{Scatter graph of 25,000 chromosomes by superphylum, GC skew versus TA skew}
\label{fig:gc-ta-scatter}
\end{figure}


\subsubsection*{Firmicute prediction}
In many bacteria, genes tend to concentrate on the leading replication strand. If the codon bias of genes is such that they are relatively rich in one nucleotide, the ``strand bias'' may itself give rise to GC or TA bias. Or in other words, if genes contain a lot of G's and they huddle on the leading strand, that strand will show GC skew. As an hypothesis, we can plot the observed GC and TA skews for all Firmicutes for which we have data.

\begin{figure}[ht]
  \centering
    \begin{minipage}[b]{0.45\linewidth}
      \includegraphics[width=\linewidth]{firmi.pdf}
      \caption{Predicted versus actual GC/TA skew for 4093 Firmicutes}
      \label{fig:the-big-graph}
    \end{minipage}
    \quad
    \begin{minipage}[b]{0.45\linewidth}
      \includegraphics[width=\linewidth]{cdif-histo.pdf}
      \caption{Scatter graph of codon/strand bias versus GC/TA skew for \emph{C. difficile}}
      \label{fig:cdif-scatter}
    \end{minipage}
\end{figure}


Mathematically the relation between the codon bias, strand bias and predicted GC skew turns out to be a simple multiplication. In figure \ref{fig:the-big-graph}, the x-axis represents this multiplication. The y-axis represents the GC and TA skew ratio. 

It can clearly be seen that both GC and TA skew correlate strongly with the codon/strand bias product. TA skew goes to zero with the two biases, but GC skew appears to persist even in the absence of such biases.

Figure \ref{fig:cdif-scatter} shows the situation within an individual chromosome (\emph{C. difficile}), based on overlapping 40960-nucleotide segments. On the x-axis we find the strand bias for these segments, running from entirely negative sense genes to entirely positive sense genes. The skew is meanwhile plotted on the y-axis, and here too we see that TA skew goes to zero in the absence of strand bias, while GC skew persists and clearly has an independent strand-based component.

\subsubsection*{Asymmetric skew}
The vast majority of chromosomes show similar skews on the leading and the lagging replication strands, something that makes sense given the pairing rules. There are however many chromosomes that have very asymmetric skews, with one strand sometimes showing no skew at all. In figure \ref{fig:asym-skew} four chromosomes are shown that exhibit such behavior. The \emph{Skew}DB lists around 250 chromosomes where one strand has a GC skew at least 3 times bigger/smaller than the other one.

\begin{figure}[ht]
  \centering
  \begin{minipage}[b]{0.45\linewidth}
    \includegraphics[width=\linewidth]{flat-skew.pdf}
    \caption{Chromosomes with asymmetric skews}
    \label{fig:asym-skew}
  \end{minipage}
  \quad
  \begin{minipage}[b]{0.45\linewidth}
    \includegraphics[width=\linewidth]{strand-div.pdf}
    \caption{Chromosomes with differing strand lengths}
    \label{fig:strand-div}
  \end{minipage}
\end{figure}

\subsubsection*{Asymmetric strands}
Bacteria tend to have very equally sized replication strands, which is also an optimum for the duration of replication. It is therefore interesting to observe that GC skew analysis finds many chromosomes where one strand is four times larger than the other strand.  In  figure \ref{fig:strand-div} four such chromosomes are shown. The \emph{Skew}DB lists around 100 chromosomes where one strand is at least three times as large as the other strand.


\subsubsection*{Anomalies}
% CHANGE: Cite lu_skewit_2020 to acknowledge the origin of the idea of using GC skew to detect mis-assemblies
In many ways, GC skew is like a forensic record of the historical developments in a chromosome. Horizontal gene transfer, inversions, integration of plasmids, excisions can all leave traces. In addition, DNA sequencing or assembly artifacts will also reliably show up in GC graphs, as elucidated with examples in \cite{lu_skewit_2020}.

Figure \ref{fig:anomalous} shows GC and TA skews for \emph{Salmonella enterica subsp. enterica serovar Concord} strain AR-0407 (NZ\_CP044177.1), and many things could be going on here. The peaks might correspond to multiple origins of replication, but might also indicate inversions or DNA assembly problems.

\begin{figure}[ht]
\centering
\includegraphics[width=.5\linewidth]{anomalous.pdf}
% CHANGE: Add italics to species name
\caption{GC and TA skew for \emph{Salmonella enterica subsp. enterica serovar Concord} strain AR-0407}
\label{fig:anomalous}
\end{figure}

\section*{Data Records}
% CHANGE: Add proper citation for the dataset and the software
The \emph{Skew}DB is available through \url{https://skewdb.org}, where it is frequently (\& automatically) refreshed. A snapshot of the data has also been deposited on Dryad\cite{https://doi.org/10.5061/dryad.g4f4qrfr6}.

The \emph{Skew}DB consists of several CSV files: skplot.csv, results.csv, genomes.csv and codongc.csv.
In addition, for each chromosome or plasmid, a separate \_fit.csv file is generated, which contains data at 4096-nucleotide resolution.

skplot.csv contains all the 4096-nucleotide resolution data as one big file for all processed chromosomes and plasmids. The parameters are described in table \ref{tab:skplot}.
\begin{table}[ht]
\begin{tabular}{|l|l|l|l|}
\hline
abspos     & locus in chromosome                       & name      & RefSeq ID                                   \\ \hline
acounts0-4 & A nucleotide counter                      & ngcount   & Counter of non-coding nucleotides           \\ \hline
ccounts0-4 & C nucleotide counter                      & pospos    & cumulative positive sense nucleotide counter \\ \hline
gcounts0-4 & G nucleotide counter                      & relpos    & relative position within chromosome/plasmid \\ \hline
tcounts0-4 & T nucleotide counter                      & taskew    & cumulative TA skew                          \\ \hline
gcskew     & cumulative GC skew                        & taskew0-3 & cumulative TA skew per codon position       \\ \hline
gcskew0-3  & cumulative GC skew per codon position     & taskewNG  & cumulative TA skew for non-coding regions   \\ \hline
gcskewNG   & cumulative GC skew for non-coding regions &           &                                             \\ \hline
\end{tabular}
\caption{Fields of skplot.csv}
\label{tab:skplot}
\end{table}

results.csv meanwhile contains the details of the fits. In this table \ref{tab:skew-table}, all marked out squares exist. The actual fields
are called alpha1gc, alpha2gc, gcRMS, alpha1ta, alpha2ta etc. DNA sequence shift and div are also specified, and they come from the GC skew.
gc0-2, ta0-2 refers to codon position. gcng and tang refer to the non-coding region skews. Finally sb denotes the strand bias. 

\begin{table}[H]
  \centering
\begin{tabular}{|l|l|l|l|ll}
\hline
     & alpha1 & alpha2 & rms & \multicolumn{1}{l|}{div} & \multicolumn{1}{l|}{shift} \\ \hline
gc   & X      &  X     &  X  & \multicolumn{1}{l|}{X}   & \multicolumn{1}{l|}{X}     \\ \hline
ta   & X      &  X     &  X  &                          &                            \\ \cline{1-4}
gc0  & X      &  X     &  X  &                          &                            \\ \cline{1-4}
gc1  & X      &  X     &  X  &                          &                            \\ \cline{1-4}
gc2  & X      &  X     &  X  &                          &                            \\ \cline{1-4}
ta0  & X      &  X     &  X  &                          &                            \\ \cline{1-4}
ta1  & X      &  X     &  X  &                          &                            \\ \cline{1-4}
ta2  & X      &  X     &  X  &                          &                            \\ \cline{1-4}
gcng & X      &  X     &  X  &                          &                            \\ \cline{1-4}
tang & X      &  X     &  X  &                          &                            \\ \cline{1-4}
sb   & X      &  X     &  X  &                          &                            \\ \cline{1-4}
\end{tabular}
\caption{Skew metrics}
\label{tab:skew-table}
\end{table}

Table \ref{tab:codongc-table} documents the data on codon bias, also split out by leading or lagging strand found in codongc.csv.
\begin{table}[H]
\begin{tabular}{|l|l|}
\hline
afrac, cfrac, gfrac, tfrac                 & Fraction of coding nucleotides that are A, C, G or T                \\ \hline
leadafrac, leadcfrac, leadgfrac, leadtfrac & Fraction of leading strand coding nucleotides that are A, C, G or T \\ \hline
lagafrac, lagcfrac, laggfrac, lagtfrac     & Fraction of lagging strand coding nucleotides that are A, C, G or T \\ \hline
ggcfrac, cgcfrac                           & The G and C fraction of GC coding nucleotides respectively          \\ \hline
atafrac, ttafrac                           & The A and T fraction of AT coding nucleotides respectively          \\ \hline
\end{tabular}
\caption{Fields in codongc.csv}
\label{tab:codongc-table}
\end{table}

The genomes.csv file contains the following fields:

\begin{table}[H]
\begin{tabular}{|l|l|}
  \hline
fullname                 & The full chromosome name as found in the FASTA file                \\ \hline
acount, ccount, gcount, tcount                 & Count of A, C, G or T nucleotides                \\ \hline
plasmid & Set to 1 in case this sequence is a plasmid \\ \hline
realm1-5     & NCBI sourced taxonomic data \\ \hline
protgenecount                           & Number of protein coding genes processed          \\ \hline
stopTAG, TAA, TGA & Number of TAG, TAA and TGA stop codons respectively          \\ \hline
stopXXX & Number of anomalous stop codons  \\ \hline
startATG, GTG, TTG & Number of ATG, GTG and TTG start codons respectively          \\ \hline
startXXX & Number of unusual start codons  \\ \hline
dnaApos & position of DnaA gene (not DnaA box!) in the DNA sequence. -1 if not found.  \\ \hline
\end{tabular}
\caption{Fields in genomes.csv}
\label{tab:genomes-table}
\end{table}


Finally, the individual \_fit.csv files contain fields called ``Xskew'' and ``predXskew'' to denote the observed X=gc, ta etc skew, plus the prediction based on the parameters found in results.csv. 
\section*{Technical Validation}
This database models the skews of many chromosomes and plasmids. Validation consists of evaluating the goodness-of-fit compared to the directly available numbers.

The \emph{Skew}DB fits skews to a relatively simple model of only four parameters. This prevents overfitting, and this model has proven to be robust in practice. Yet, when doing automated analysis of tens of thousands of chromosomes, mistakes will be made. Also, not all organisms show coherent GC skew.

\begin{figure}[tbhp]
\centering
\includegraphics[width=.5\linewidth]{rms-samples.pdf}
\caption{\emph{Skew}DB fits for 16 equal sized quality categories of bacterial chromosomes}
\label{fig:rms-samples}
\end{figure}

Figure \ref{fig:rms-samples} shows 16 equal sized quality categories, where it is visually clear that the 88\% best fits are excellent. It is therefore reasonable to filter the database on $RMS_{gc}<0.16$. Or conversely, it could be said that above this limit interesting anomalous chromosomes can be found. 

The DoriC database \cite{luo_doric_2019} contains precise details of the location of the origin of replication. 2267 sequences appear both in DoriC and in the \emph{Skew}DB. The DoriC origin of replication should roughly be matched by the ``shift'' metric in the \emph{Skew}DB (but see Usage notes). For 90\% of sequences appearing in both databases, there is less than 5\% relative chromosome distance between these two independent metrics. This is encouraging since these two numbers do not quite measure the same thing.

On a similar note, the DnaA gene is typically (but not necessarily) located near the origin of replication. For over 80\% of chromosomes, DnaA is found within 5\% of the \emph{Skew}DB ``shift'' metric. This too is an encouraging independent confirmation of the accuracy of the data.

Finally, during processing numbers are kept of the start and stop codons encountered on all protein coding genes on all chromosomes and plasmids. These numbers are interesting in themselves (because they correlate with GC content, for example), but they also match published frequencies, and show limited numbers of non-canonical start codons, and around 0.005\% anomalous stop codons. This too confirms that the analyses are based on correct (annotation) assumptions.


\section*{Usage Notes}
The existential limitation of any database like the \emph{Skew}DB is that it does not represent the distribution of organisms found in nature. The sequence and annotation databases are dominated by easily culturable microbes. And even within that selection, specific (model) organisms are heavily oversampled because of their scientific, economic or medical relevance.

Because of this, care should be taken to interpret numbers in a way that takes such over- and undersampling into account. This leaves enough room however for finding correlations. Some metrics are sampled so heavily that it would be a miracle if the unculturable organisms were collectively conspiring to skew the statistics away from the average. In addition, the database is a very suitable way to test or generate hypotheses, or to find anomalous organisms.

Finally it should be noted that the \emph{Skew}DB tries to precisely measure the skew parameters, but it makes no effort to pin down the Origin of replication exactly. For such uses, please refer to the DoriC database \cite{luo_doric_2019}. In future work, the \emph{Skew}DB will attempt to use OriC motifs to improve fitting of this metric.

On \url{https://skewdb.org} an explanatory Jupyter \cite{Kluyver:2016aa} notebook can be found that uses Matplotlib \cite{Hunter:2007} and Pandas \cite{jeff_reback_2021_5203279} to create all the graphs from this article, and many more. In addition, this notebook reproduces all numerical claims made in this work. The \emph{Skew}DB website also provides links to informal articles that further explain GC skew, and how it could be used for research.

\section*{Code availability}
The \emph{Skew}DB is produced using the \href{https://github.com/berthubert/antonie2}{Antonie DNA processing software}, which is open source. In addition the pipeline is fully automated and reproducible, including the retrieval of sequences, annotations and taxonomic data from the NCBI website. The software has also been deposited with Zenodo\cite{https://doi.org/10.5281/zenodo.5516524}.

A \href{https://github.com/berthubert/skewdb-articles}{GitHub repository} is available for this article, which includes this reproducible pipeline, plus a script that regenerates all the graphs and numerical claims from this paper. 

\begin{thebibliography}{10}
\urlstyle{rm}
\expandafter\ifx\csname url\endcsname\relax
  \def\url#1{\texttt{#1}}\fi
\expandafter\ifx\csname urlprefix\endcsname\relax\def\urlprefix{URL }\fi
\expandafter\ifx\csname doiprefix\endcsname\relax\def\doiprefix{DOI: }\fi
\providecommand{\bibinfo}[2]{#2}
\providecommand{\eprint}[2][]{\url{#2}}

\bibitem{pmid:10570985}
\bibinfo{author}{Frank, A.~C.} \& \bibinfo{author}{Lobry, J.~R.}
\newblock \bibinfo{journal}{\bibinfo{title}{Asymmetric substitution patterns: a
  review of possible underlying mutational or selective mechanisms.}}
\newblock {\emph{\JournalTitle{Gene}}} \textbf{\bibinfo{volume}{238}},
  \bibinfo{pages}{65--77} (\bibinfo{year}{1999}).

\bibitem{Marn2008}
\bibinfo{author}{Mar{\'{\i}}n, A.} \& \bibinfo{author}{Xia, X.}
\newblock \bibinfo{journal}{\bibinfo{title}{{GC} skew in protein-coding genes
  between the leading and lagging strands in bacterial genomes: New
  substitution models incorporating strand bias}}.
\newblock {\emph{\JournalTitle{Journal of Theoretical Biology}}}
  \textbf{\bibinfo{volume}{253}}, \bibinfo{pages}{508--513},
  \url{https://doi.org/10.1016/j.jtbi.2008.04.004} (\bibinfo{year}{2008}).

\bibitem{Quan2019}
\bibinfo{author}{Quan, C.-L.} \& \bibinfo{author}{Gao, F.}
\newblock \bibinfo{journal}{\bibinfo{title}{Quantitative analysis and
  assessment of base composition asymmetry and gene orientation bias in
  bacterial genomes}}.
\newblock {\emph{\JournalTitle{{FEBS} Letters}}}
  \textbf{\bibinfo{volume}{593}}, \bibinfo{pages}{918--925},
  \url{https://doi.org/10.1002/1873-3468.13374} (\bibinfo{year}{2019}).

\bibitem{lu_skewit_2020}
\bibinfo{author}{Lu, J.} \& \bibinfo{author}{Salzberg, S.~L.}
\newblock \bibinfo{journal}{\bibinfo{title}{{SkewIT}: {The} {Skew} {Index}
  {Test} for large-scale {GC} {Skew} analysis of bacterial genomes}}.
\newblock {\emph{\JournalTitle{PLOS Computational Biology}}}
  \textbf{\bibinfo{volume}{16}}, \bibinfo{pages}{e1008439},
  \url{https://doi.org/10.1371/journal.pcbi.1008439} (\bibinfo{year}{2020}).

\bibitem{luo_doric_2019}
\bibinfo{author}{Luo, H.} \& \bibinfo{author}{Gao, F.}
\newblock \bibinfo{journal}{\bibinfo{title}{{DoriC} 10.0: an updated database
  of replication origins in prokaryotic genomes including chromosomes and
  plasmids}}.
\newblock {\emph{\JournalTitle{Nucleic Acids Research}}}
  \textbf{\bibinfo{volume}{47}}, \bibinfo{pages}{D74--D77},
  \url{https://doi.org/10.1093/nar/gky1014} (\bibinfo{year}{2019}).

\bibitem{ODonnell2013}
\bibinfo{author}{O{\textquotesingle}Donnell, M.}, \bibinfo{author}{Langston,
  L.} \& \bibinfo{author}{Stillman, B.}
\newblock \bibinfo{journal}{\bibinfo{title}{Principles and concepts of {DNA}
  replication in bacteria, archaea, and eukarya}}.
\newblock {\emph{\JournalTitle{Cold Spring Harbor Perspectives in Biology}}}
  \textbf{\bibinfo{volume}{5}}, \bibinfo{pages}{a010108--a010108},
  \url{https://doi.org/10.1101/cshperspect.a010108} (\bibinfo{year}{2013}).

\bibitem{Lilly2015}
\bibinfo{author}{Lilly, J.} \& \bibinfo{author}{Camps, M.}
\newblock \bibinfo{journal}{\bibinfo{title}{Mechanisms of theta plasmid
  replication}}.
\newblock {\emph{\JournalTitle{Microbiology Spectrum}}}
  \textbf{\bibinfo{volume}{3}}, \url{https://doi.org/10.1128/microbiolspec.plas-0029-2014}
  (\bibinfo{year}{2015}).

\bibitem{rudner_separation_1968}
\bibinfo{author}{Rudner, R.}, \bibinfo{author}{Karkas, J.~D.} \&
  \bibinfo{author}{Chargaff, E.}
\newblock \bibinfo{journal}{\bibinfo{title}{Separation of {B}. subtilis {DNA}
  into complementary strands. 3. {Direct} analysis.}}
\newblock {\emph{\JournalTitle{Proceedings of the National Academy of
  Sciences}}} \textbf{\bibinfo{volume}{60}}, \bibinfo{pages}{921--922},
  \url{https://doi.org/10.1073/pnas.60.3.921} (\bibinfo{year}{1968}).

\bibitem{fariselli_dna_2020}
\bibinfo{author}{Fariselli, P.}, \bibinfo{author}{Taccioli, C.},
  \bibinfo{author}{Pagani, L.} \& \bibinfo{author}{Maritan, A.}
\newblock \bibinfo{journal}{\bibinfo{title}{{DNA} sequence symmetries from
  randomness: the origin of the {Chargaff}’s second parity rule}}.
\newblock {\emph{\JournalTitle{Briefings in Bioinformatics}}}
  \bibinfo{pages}{bbaa041}, \url{https://doi.org/10.1093/bib/bbaa041} (\bibinfo{year}{2020}).

\bibitem{tillier_contributions_2000}
\bibinfo{author}{Tillier, E.~R.} \& \bibinfo{author}{Collins, R.~A.}
\newblock \bibinfo{journal}{\bibinfo{title}{The {Contributions} of
  {Replication} {Orientation}, {Gene} {Direction}, and {Signal} {Sequences} to
  {Base}-{Composition} {Asymmetries} in {Bacterial} {Genomes}}}.
\newblock {\emph{\JournalTitle{Journal of Molecular Evolution}}}
  \textbf{\bibinfo{volume}{50}}, \bibinfo{pages}{249--257},
  \url{https://doi.org/10.1007/s002399910029} (\bibinfo{year}{2000}).

\bibitem{zhang_brief_2014}
\bibinfo{author}{Zhang, R.} \& \bibinfo{author}{Zhang, C.-T.}
\newblock \bibinfo{journal}{\bibinfo{title}{A {Brief} {Review}: {The} {Z}-curve
  {Theory} and its {Application} in {Genome} {Analysis}}}.
\newblock {\emph{\JournalTitle{Current genomics}}}
  \textbf{\bibinfo{volume}{15}}, \bibinfo{pages}{78--94},
  \url{https://doi.org/10.2174/1389202915999140328162433} (\bibinfo{year}{2014}).
\newblock \bibinfo{note}{Publisher: Bentham Science Publishers}.

\bibitem{charneski_atypical_2011}
\bibinfo{author}{Charneski, C.~A.}, \bibinfo{author}{Honti, F.},
  \bibinfo{author}{Bryant, J.~M.}, \bibinfo{author}{Hurst, L.~D.} \&
  \bibinfo{author}{Feil, E.~J.}
\newblock \bibinfo{journal}{\bibinfo{title}{Atypical {AT} {Skew} in {Firmicute}
  {Genomes} {Results} from {Selection} and {Not} from {Mutation}}}.
\newblock {\emph{\JournalTitle{PLOS Genetics}}} \textbf{\bibinfo{volume}{7}},
  \bibinfo{pages}{e1002283}, \url{https://doi.org/10.1371/journal.pgen.1002283}
  (\bibinfo{year}{2011}).

\bibitem{grigoriev_analyzing_1998}
\bibinfo{author}{Grigoriev, A.}
\newblock \bibinfo{journal}{\bibinfo{title}{Analyzing genomes with cumulative
  skew diagrams}}.
\newblock {\emph{\JournalTitle{Nucleic Acids Research}}}
  \textbf{\bibinfo{volume}{26}}, \bibinfo{pages}{2286--2290},
  \url{https://doi.org/10.1093/nar/26.10.2286} (\bibinfo{year}{1998}).

\bibitem{Roten2002}
\bibinfo{author}{Roten, C.-A.~H.}
\newblock \bibinfo{journal}{\bibinfo{title}{Comparative genometrics ({CG}): a
  database dedicated to biometric comparisons of whole genomes}}.
\newblock {\emph{\JournalTitle{Nucleic Acids Research}}}
  \textbf{\bibinfo{volume}{30}}, \bibinfo{pages}{142--144},
  \url{https://doi.org/10.1093/nar/30.1.142} (\bibinfo{year}{2002}).

\bibitem{Zhang2003}
\bibinfo{author}{Zhang, C.-T.}, \bibinfo{author}{Zhang, R.} \&
  \bibinfo{author}{Ou, H.-Y.}
\newblock \bibinfo{journal}{\bibinfo{title}{The z curve database: a graphic
  representation of genome sequences}}.
\newblock {\emph{\JournalTitle{Bioinformatics}}} \textbf{\bibinfo{volume}{19}},
  \bibinfo{pages}{593--599}, \url{https://doi.org/10.1093/bioinformatics/btg041}
  (\bibinfo{year}{2003}).

\bibitem{Thomas2007}
\bibinfo{author}{Thomas, J.~M.}, \bibinfo{author}{Horspool, D.},
  \bibinfo{author}{Brown, G.}, \bibinfo{author}{Tcherepanov, V.} \&
  \bibinfo{author}{Upton, C.}
\newblock \bibinfo{journal}{\bibinfo{title}{{GraphDNA}: a java program for
  graphical display of {DNA} composition analyses}}.
\newblock {\emph{\JournalTitle{{BMC} Bioinformatics}}}
  \textbf{\bibinfo{volume}{8}}, \url{https://doi.org/10.1186/1471-2105-8-21}
  (\bibinfo{year}{2007}).

\bibitem{Grigoriev1998}
\bibinfo{author}{Grigoriev, A.}
\newblock \bibinfo{journal}{\bibinfo{title}{Analyzing genomes with cumulative
  skew diagrams}}.
\newblock {\emph{\JournalTitle{Nucleic Acids Research}}}
  \textbf{\bibinfo{volume}{26}}, \bibinfo{pages}{2286--2290},
  \url{https://doi.org/10.1093/nar/26.10.2286} (\bibinfo{year}{1998}).

\bibitem{Nelder1965}
\bibinfo{author}{Nelder, J.~A.} \& \bibinfo{author}{Mead, R.}
\newblock \bibinfo{journal}{\bibinfo{title}{A simplex method for function
  minimization}}.
\newblock {\emph{\JournalTitle{The Computer Journal}}}
  \textbf{\bibinfo{volume}{7}}, \bibinfo{pages}{308--313},
  \url{https://doi.org/10.1093/comjnl/7.4.308} (\bibinfo{year}{1965}).

\bibitem{https://doi.org/10.5061/dryad.g4f4qrfr6}
\bibinfo{author}{Hubert, B.}
\newblock \bibinfo{title}{Skewdb: A comprehensive database of gc and 10 other
  skews for over 28, 000 chromosomes and plasmids},
  \url{https://doi.org/10.5061/DRYAD.G4F4QRFR6} (\bibinfo{year}{2021}).

\bibitem{Kluyver:2016aa}
\bibinfo{author}{Kluyver, T.} \emph{et~al.}
\newblock \bibinfo{title}{Jupyter notebooks -- a publishing format for
  reproducible computational workflows}.
\newblock In \bibinfo{editor}{Loizides, F.} \& \bibinfo{editor}{Schmidt, B.}
  (eds.) \emph{\bibinfo{booktitle}{Positioning and Power in Academic
  Publishing: Players, Agents and Agendas}}, \bibinfo{pages}{87 -- 90}
  (\bibinfo{organization}{IOS Press}, \bibinfo{year}{2016}).

\bibitem{Hunter:2007}
\bibinfo{author}{Hunter, J.~D.}
\newblock \bibinfo{journal}{\bibinfo{title}{Matplotlib: A 2d graphics
  environment}}.
\newblock {\emph{\JournalTitle{Computing in Science \& Engineering}}}
  \textbf{\bibinfo{volume}{9}}, \bibinfo{pages}{90--95},
  \url{https://doi.org/10.1109/MCSE.2007.55} (\bibinfo{year}{2007}).

\bibitem{jeff_reback_2021_5203279}
\bibinfo{author}{Reback, J.} \emph{et~al.}
\newblock \bibinfo{title}{pandas-dev/pandas: Pandas 1.3.2},
  \url{https://doi.org/10.5281/zenodo.5203279} (\bibinfo{year}{2021}).

\bibitem{https://doi.org/10.5281/zenodo.5516524}
\bibinfo{author}{Hubert, B.} \& \bibinfo{author}{{Beaumont Lab}}.
\newblock \bibinfo{title}{berthubert/antonie2: Skewversion 1.0},
  \url{https://doi.org/10.5281/ZENODO.5516524} (\bibinfo{year}{2021}).

\bibitem{hol_density-dependent_2016}
\bibinfo{author}{Hol, F. J.~H.}, \bibinfo{author}{Hubert, B.},
  \bibinfo{author}{Dekker, C.} \& \bibinfo{author}{Keymer, J.~E.}
\newblock \bibinfo{journal}{\bibinfo{title}{Density-dependent adaptive
  resistance allows swimming bacteria to colonize an antibiotic gradient}}.
\newblock {\emph{\JournalTitle{The ISME Journal}}}
  \textbf{\bibinfo{volume}{10}}, \bibinfo{pages}{30--38},
  \url{https://doi.org/10.1038/ismej.2015.107} (\bibinfo{year}{2016}).

\end{thebibliography}


\section*{Acknowledgements} 

I would like to thank Bertus Beaumont for helping me to think like a biologist, and Jason Piper for regularly pointing me to the relevant literature. In addition, I am grateful that Felix Hol kindly allowed me to field test my software on his DNA sequences \cite{hol_density-dependent_2016}. Twitter users $@$halvorz and $@$Suddenly\_a\_goat also provided valuable feedback.

\section*{Author contributions statement}

B.H. did all the work.

\section*{Competing interests}
The author declares no competing interests.

\section*{Figures \& Tables}
\newcommand{\comment}[1]{}

\comment{
}

\end{document}
